src/Language/PureScript/Types.hs

```diff
@@ -516,39 +516,48 @@ isMonoType _        = True
 mkForAll :: [(a, (Text, Maybe (Type a)))] -> Type a -> Type a
 mkForAll args ty = foldr (\(ann, (arg, mbK)) t -> ForAll ann TypeVarInvisible arg mbK t Nothing) ty args
 
+-- | Rename a type variable, taking into account variable shadowing
+renameTypeVar :: Text -> Text -> Type a -> Type a
+renameTypeVar original replacement = replaceAllTypeVars' [(original, Left replacement)]
+
 -- | Replace a type variable, taking into account variable shadowing
 replaceTypeVars :: Text -> Type a -> Type a -> Type a
 replaceTypeVars v r = replaceAllTypeVars [(v, r)]
 
 -- | Replace named type variables with types
 replaceAllTypeVars :: [(Text, Type a)] -> Type a -> Type a
-replaceAllTypeVars = go [] where
-  go :: [Text] -> [(Text, Type a)] -> Type a -> Type a
-  go _  m (TypeVar ann v) = fromMaybe (TypeVar ann v) (v `lookup` m)
+replaceAllTypeVars rs = replaceAllTypeVars' (fmap (fmap Right) rs)
```

Comment:
I'm confused about all the changes in this file; what are they for?

Reply:
That fixes #4495.

Reply:
That fix is included in this PR because it only becomes noticeable once we make the changes in this PR.

Reply:
Ah, okay. Then I have notes.